src/or/control.c

```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CIRCUIT_ENDED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CONNECTION_ENDED)) {
+    return 0;
+  }
+
+  /* Ignore events which are missing both connection and circuit */
+  if (!conn && !circ) {
+    return 0;
+  }
+
+  if (privcount_is_overhead(conn, circ)) {
+    return 0;
+  }
+
+  /* Cells get counted at non-exit connections as well as exit connections */
+  return 1;
+}
+
+/* Should PrivCount ignore DNS requests from this connection or circuit?
+ */
+int
+privcount_is_counted_for_dns(const edge_connection_t* exitconn,
+                             const or_circuit_t *orcirc)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_DNS_RESOLVED)) {
+    return 0;
+  }
+
+  /* Ignore events with incomplete data */
+  if (!exitconn || !orcirc) {
+    return 0;
+  }
+
+  /* DNS connections are never directory connections */
+  int is_dir = privcount_is_dir(TO_CONN(exitconn), TO_CIRCUIT(orcirc));
+  tor_assert_nonfatal(!is_dir);
+
+  /* DNS is only available at exit connections */
+  return privcount_is_exit(exitconn, orcirc);
+}
+
+/* Is the remote end of chan a relay in our current consensus? */
+static int
+privcount_is_consensus_relay(const channel_t* chan)
+{
+  if (!chan) {
+    /* Assume that missing connections *aren't* a known relay */
+    return 0;
+  }
+
+  if (connection_or_digest_is_known_relay(chan->identity_digest)) {
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+/* Is the remote end of chan a client or bridge?
+ * (Relays authenticate their peer IDs to us, clients and bridges do not) */
+static int
+privcount_is_unauthenticated_client(const channel_t *chan)
+{
+  if (!chan) {
+    /* Assume that missing connections *aren't* a client */
+    return 0;
+  }
+
+  /* channel_is_client doesn't actually modify chan */
+  if (channel_is_client((channel_t *)chan)) {
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+/* Is the remote end of chan a client or bridge? */
+static int
+privcount_is_client(const channel_t* chan)
+{
+  int is_relay = privcount_is_consensus_relay(chan);
+  int is_client = privcount_is_unauthenticated_client(chan);
+
+  /* Relays in the consensus should always authenticate their peer IDs with us
+   */
+  if (is_relay) {
+    tor_assert_nonfatal(!is_client);
+  }
+
+  /* is_client is more reliable, but only with the fix for Tor bug 21406 */
+  return is_client && !is_relay;
+}
+
+/* Perform a 64-bit saturating add of a and b */
+static uint64_t
+privcount_add_saturating(uint64_t a, uint64_t b) {
+  if (PREDICT_LIKELY(UINT64_MAX - a > b)) {
+    return a + b;
+  } else {
+    return UINT64_MAX;
+  }
+}
+
+/* Perform a += b without overflow, and without returning a value */
+void
+privcount_sum(uint64_t *total, uint64_t increment) {
```

Comment:
I'm not sure why this is safer/cleaner than just calling `privcount_add_saturating` directly from wherever `privcount_sum` is currently called. I realize it is a bit shorter because you don't have to do the assignment in the caller. Maybe it's just programmer style, but I usually avoid passing pointers and then operating on them when unnecessary (like when passing ints).